From ccaf4694673662d3f7a064ee3a2b01920c5231a8 Mon Sep 17 00:00:00 2001
From: Honggyu Kim <honggyu.kim@sk.com>
Date: Fri, 5 Apr 2024 15:08:56 +0900
Subject: [PATCH] mm/damon: Add "damon_migrate_{hot,cold}" vmstat

This patch adds "damon_migrate_{hot,cold}" under node specific vmstat
counters at the following location.

  /sys/devices/system/node/node*/vmstat

The counted values are accumulcated to the global vmstat so it also
introduces the same counter at /proc/vmstat as well.

Signed-off-by: Honggyu Kim <honggyu.kim@sk.com>
Link: https://lore.kernel.org/20240405060858.2818-8-honggyu.kim@sk.com
Signed-off-by: SeongJae Park <sj@kernel.org>
---
 include/linux/mmzone.h |  4 ++++
 mm/damon/paddr.c       | 30 ++++++++++++++++++++++++------
 mm/vmstat.c            |  4 ++++
 3 files changed, 32 insertions(+), 6 deletions(-)

diff --git a/include/linux/mmzone.h b/include/linux/mmzone.h
index c11b7cde81ef..42eafd3e3232 100644
--- a/include/linux/mmzone.h
+++ b/include/linux/mmzone.h
@@ -217,6 +217,10 @@ enum node_stat_item {
 	PGDEMOTE_KSWAPD,
 	PGDEMOTE_DIRECT,
 	PGDEMOTE_KHUGEPAGED,
+#ifdef CONFIG_DAMON_PADDR
+	DAMON_MIGRATE_HOT,
+	DAMON_MIGRATE_COLD,
+#endif
 	NR_VM_NODE_STAT_ITEMS
 };
 
diff --git a/mm/damon/paddr.c b/mm/damon/paddr.c
index 8182af8cb950..a0cc48ed3e63 100644
--- a/mm/damon/paddr.c
+++ b/mm/damon/paddr.c
@@ -251,10 +251,23 @@ enum migration_mode {
 
 static unsigned int __damon_pa_migrate_folio_list(
 		struct list_head *migrate_folios, struct pglist_data *pgdat,
-		int target_nid)
+		enum migration_mode mode, int target_nid)
 {
 	unsigned int nr_succeeded;
 	nodemask_t allowed_mask = NODE_MASK_NONE;
+	enum node_stat_item node_stat;
+
+	switch (mode) {
+	case MIG_MIGRATE_HOT:
+		node_stat = DAMON_MIGRATE_HOT;
+		break;
+	case MIG_MIGRATE_COLD:
+		node_stat = DAMON_MIGRATE_COLD;
+		break;
+	default:
+		return 0;
+	}
+
 	struct migration_target_control mtc = {
 		/*
 		 * Allocate from 'node', or fail quickly and quietly.
@@ -278,11 +291,14 @@ static unsigned int __damon_pa_migrate_folio_list(
 		      (unsigned long)&mtc, MIGRATE_ASYNC, MR_DAMON,
 		      &nr_succeeded);
 
+	mod_node_page_state(pgdat, node_stat, nr_succeeded);
+
 	return nr_succeeded;
 }
 
 static unsigned int damon_pa_migrate_folio_list(struct list_head *folio_list,
 						struct pglist_data *pgdat,
+						enum migration_mode mode,
 						int target_nid)
 {
 	unsigned int nr_migrated = 0;
@@ -312,7 +328,7 @@ static unsigned int damon_pa_migrate_folio_list(struct list_head *folio_list,
 
 	/* Migrate folios selected for migration */
 	nr_migrated += __damon_pa_migrate_folio_list(
-			&migrate_folios, pgdat, target_nid);
+			&migrate_folios, pgdat, mode, target_nid);
 	/*
 	 * Folios that could not be migrated are still in @migrate_folios.  Add
 	 * those back on @folio_list
@@ -334,7 +350,8 @@ static unsigned int damon_pa_migrate_folio_list(struct list_head *folio_list,
 }
 
 static unsigned long damon_pa_migrate_pages(struct list_head *folio_list,
-					    int target_nid)
+					    int target_nid,
+					    enum migration_mode mode)
 {
 	int nid;
 	unsigned long nr_migrated = 0;
@@ -357,13 +374,13 @@ static unsigned long damon_pa_migrate_pages(struct list_head *folio_list,
 
 		nr_migrated += damon_pa_migrate_folio_list(&node_folio_list,
 							   NODE_DATA(nid),
-							   target_nid);
+							   mode, target_nid);
 		nid = folio_nid(lru_to_folio(folio_list));
 	} while (!list_empty(folio_list));
 
 	nr_migrated += damon_pa_migrate_folio_list(&node_folio_list,
 						   NODE_DATA(nid),
-						   target_nid);
+						   mode, target_nid);
 
 	memalloc_noreclaim_restore(noreclaim_flag);
 
@@ -422,7 +439,8 @@ static unsigned long damon_pa_migrate(struct damon_region *r, struct damos *s,
 		break;
 	case MIG_MIGRATE_HOT:
 	case MIG_MIGRATE_COLD:
-		applied = damon_pa_migrate_pages(&folio_list, s->target_nid);
+		applied = damon_pa_migrate_pages(&folio_list, s->target_nid,
+				mode);
 		break;
 	default:
 		/* Unexpected migration mode. */
diff --git a/mm/vmstat.c b/mm/vmstat.c
index db79935e4a54..be9ba89fede1 100644
--- a/mm/vmstat.c
+++ b/mm/vmstat.c
@@ -1252,6 +1252,10 @@ const char * const vmstat_text[] = {
 	"pgdemote_kswapd",
 	"pgdemote_direct",
 	"pgdemote_khugepaged",
+#ifdef CONFIG_DAMON_PADDR
+	"damon_migrate_hot",
+	"damon_migrate_cold",
+#endif
 
 	/* enum writeback_stat_item counters */
 	"nr_dirty_threshold",
-- 
2.39.2

